File: Practica5/SeleccionDeActividades.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package practica5algoritmosvoraces;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

/**
 *
 * @author angel
 */

class Actividad {
    int inicio, fin;

    Actividad(int i, int f) {
        inicio = i;
        fin = f;
    }
}

public class SeleccionDeActividades {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Número de actividades: ");
        int n = sc.nextInt();

        List<Actividad> acts = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            System.out.print("Inicio de actividad " + (i+1) + ": ");
            int ini = sc.nextInt();
            System.out.print("Fin de actividad " + (i+1) + ": ");
            int fin = sc.nextInt();
            acts.add(new Actividad(ini, fin));
        }

        // Ordenar por tiempo de fin
        acts.sort(Comparator.comparingInt(a -> a.fin));

        System.out.println("\nActividades seleccionadas:");
        int ultimaFin = -1;

        for (Actividad a : acts) {
            if (a.inicio >= ultimaFin) {
                System.out.println("Inicio: " + a.inicio + "  Fin: " + a.fin);
                ultimaFin = a.fin;
            }
        }

        sc.close();
    }
}

File: Practica5/SeleccionDeActividades.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int inicio;
    int fin;
} Actividad;

int cmp(const void *a, const void *b) {
    Actividad *x = (Actividad*)a;
    Actividad *y = (Actividad*)b;
    return x->fin - y->fin;
}

int main() {
    int n;
    int i;

    printf("Numero de actividades: ");
    scanf("%d", &n);

    Actividad acts[n];

    for (i = 0; i < n; i++) {
        printf("Inicio de actividad %d: ", i+1);
        scanf("%d", &acts[i].inicio);
        printf("Fin de actividad %d: ", i+1);
        scanf("%d", &acts[i].fin);
    }

    // Ordenar por tiempo de fin
    qsort(acts, n, sizeof(Actividad), cmp);

    printf("\nActividades seleccionadas:\n");
    int ultimaFin = -1;

    for (i = 0; i < n; i++) {
        if (acts[i].inicio >= ultimaFin) {
            printf("Inicio: %d  Fin: %d\n", acts[i].inicio, acts[i].fin);
            ultimaFin = acts[i].fin;
        }
    }

    return 0;
}

